File: morphir/sdk/core/src/morphir/sdk/List.scala
/*
Copyright 2020 Morgan Stanley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

package morphir.sdk

import morphir.sdk.Basics.Bool
import morphir.sdk.Maybe.{ Just, Maybe }

object List {
  type List[+A] = scala.List[A]

  @inline def all[A](predicate: A => Boolean)(xs: List[A]): Boolean =
    xs.forall(predicate)

  @inline def any[A](predicate: A => Boolean)(xs: List[A]): Boolean =
    xs.exists(predicate)

  @inline def append[A](xs: List[A])(ys: List[A]): List[A] =
    xs ++ ys

  def apply[A](items: A*): List[A] =
    scala.List(items: _*)

  @inline def concat[A](lists: List[List[A]]): List[A] =
    lists.flatten

  @inline def concatMap[A, B](f: A => List[B])(lists: List[A]): List[B] =
    lists.flatMap(f)

  @inline def cons[A](head: A)(tail: List[A]): List[A] = head :: tail

  @inline def drop[A](n: Int)(xs: List[A]): List[A] = xs.drop(n)

  @inline def empty[A]: List[A] = Nil

  @inline def filter[A](f: A => Boolean)(xs: List[A]): List[A] =
    xs.filter(f)

  def filterMap[A, B](f: A => Maybe[B])(xs: List[A]): List[B] =
    xs.map(f).collect { case Just(value) => value }

  def foldl[A, B](f: A => B => B)(initial: B)(xs: List[A]): B = {
    def fn(b: B, a: A): B = f(a)(b)
    xs.foldLeft(initial)(fn)
  }

  def foldr[A, B](f: A => B => B)(initial: B)(xs: List[A]): B = {
    def fn(a: A, b: B): B = f(a)(b)
    xs.foldRight(initial)(fn)
  }

  @inline def head[A](xs: List[A]): Maybe[A] = xs.headOption

  def indexedMap[X, R](fn: Int => X => R)(xs: List[X]): List[R] =
    xs.zipWithIndex.map(tuple => fn(tuple._2)(tuple._1))

  def intersperse[A](elem: A)(xs: List[A]): List[A] = xs match {
    case lst if lst == Nil => lst
    case lst @ _ :: Nil    => lst
    case lst =>
      lst.take(xs.length - 1).flatMap(x => List(x, elem)) ++ List(xs.last)
  }

  @inline def length[A](xs: List[A]): Int    = xs.length
  @inline def singleton[A](item: A): List[A] = scala.List(item)

  @inline def map[A, B](mapping: A => B)(list: List[A]): List[B] =
    list.map(mapping)

  /** Combine two lists, combining them with the given function. If one list is longer, the extra elements are dropped.
    * @param mapping
    *   a mapping function
    * @param xs
    *   the first list
    * @param ys
    *   the second list
    * @tparam A
    *   the type of the first list
    * @tparam B
    *   the type of the second list
    * @tparam R
    *   the type of the resulting list
    * @return
    *   a list containing the combined elements of list1 and list2 using the mapping function.
    */
  def map2[A, B, R](
    mapping: A => B => R
  )(xs: List[A])(ys: List[B]): List[R] =
    xs.zip(ys).map { case (a, b) =>
      mapping(a)(b)
    }

  def map3[X, Y, Z, R](
    mapping: X => Y => Z => R
  )(xs: List[X])(ys: List[Y])(zs: List[Z]): List[R] =
    xs.zip(ys).zip(zs).map { case ((x, y), z) =>
      mapping(x)(y)(z)
    }

  def map4[A, B, C, D, R](
    mapping: A => B => C => D => R
  )(as: List[A])(bs: List[B])(cs: List[C])(ds: List[D]): List[R] =
    as.zip(bs).zip(cs).zip(ds).map { case (((a, b), c), d) =>
      mapping(a)(b)(c)(d)
    }

  def map5[A, B, C, D, E, R](
    mapping: A => B => C => D => E => R
  )(as: List[A])(bs: List[B])(cs: List[C])(ds: List[D])(es: List[E]): List[R] =
    as.zip(bs)
      .zip(cs)
      .zip(ds)
      .zip(es)
      .map { case ((((a, b), c), d), e) =>
        mapping(a)(b)(c)(d)(e)
      }

  @inline def member[A, A1 >: A](candidate: A1)(xs: List[A]): Boolean =
    xs.contains(candidate)

  @inline def isEmpty[A](list: List[A]): Boolean = list.isEmpty

  @inline def partition[A](f: A => Boolean)(xs: List[A]): (List[A], List[A]) =
    xs.partition(f)

  @inline def range(start: Int)(end: Int): List[Int] =
    scala.List.range(start, end)

  @inline def repeat[A](n: Int)(elem: => A): List[A] =
    scala.List.fill(n)(elem)

  @inline def reverse[A](xs: List[A]): List[A] = xs.reverse

  @inline def tail[A](xs: List[A]): Maybe[List[A]] = xs.tail match {
    case Nil  => Maybe.Nothing
    case tail => Maybe.Just(tail)
  }

  @inline def take[A](n: Int)(xs: List[A]): List[A] = xs.take(n)

  @inline def unzip[A, B](xs: List[(A, B)]): (List[A], List[B]) =
    xs.unzip

  @inline def minimum[A: Ordering](list: List[A]): Maybe[A] =
    if (list.isEmpty) Maybe.Nothing else Maybe.Just(list.min)
  @inline def maximum[A: Ordering](list: List[A]): Maybe[A] =
    if (list.isEmpty) Maybe.Nothing else Maybe.Just(list.max)
  @inline def sum[A: Numeric](list: List[A]): A     = list.sum
  @inline def product[A: Numeric](list: List[A]): A = list.product

  def innerJoin[A, B](listB: List[B])(f: A => B => Bool)(listA: List[A]): List[(A, B)] =
    for {
      itemA <- listA
      itemB <- listB
      if f(itemA)(itemB)
    } yield (itemA, itemB)

  def leftJoin[A, B](listB: List[B])(f: A => B => Bool)(listA: List[A]): List[(A, Maybe[B])] =
    listA.flatMap { itemA =>
      val filteredListB = listB.filter(f(itemA))
      if (filteredListB.isEmpty) {
        scala.List((itemA, Maybe.nothing))
      } else {
        for {
          itemB <- filteredListB
        } yield (itemA, Maybe.just(itemB))
      }
    }
}

File: pkg/models/ir/sdk/list.go
package sdk

import (
	ir "github.com/finos/morphir/pkg/models/ir"
)

// ListModuleName returns the module name for Morphir.SDK.List
func ListModuleName() ir.ModuleName {
	return ir.PathFromString("List")
}

// ListModuleSpec returns the module specification for Morphir.SDK.List
func ListModuleSpec() ir.ModuleSpecification[ir.Unit] {
	return ir.NewModuleSpecification(
		listTypes(),
		listValues(),
		nil,
	)
}

// listTypes returns the type specifications for the List module
func listTypes() []ir.ModuleSpecificationType[ir.Unit] {
	return []ir.ModuleSpecificationType[ir.Unit]{
		// List type (parameterized by element type)
		ir.ModuleSpecificationTypeFromParts[ir.Unit](
			ir.NameFromString("List"),
			ir.NewDocumented(
				"Type that represents a list of values.",
				ir.NewOpaqueTypeSpecification[ir.Unit]([]ir.Name{
					ir.NameFromString("a"),
				}),
			),
		),
	}
}

// listValues returns the value specifications for the List module
func listValues() []ir.ModuleSpecificationValue[ir.Unit] {
	return []ir.ModuleSpecificationValue[ir.Unit]{
		// Construction
		VSpec("singleton", []VSpecInput{
			{Name: "value", Type: TVar("a")},
		}, ListType(TVar("a")), "Create a list with only one element."),

		VSpec("repeat", []VSpecInput{
			{Name: "n", Type: intType()},
			{Name: "value", Type: TVar("a")},
		}, ListType(TVar("a")), "Create a list with n copies of a value."),

		VSpec("range", []VSpecInput{
			{Name: "start", Type: intType()},
			{Name: "end", Type: intType()},
		}, ListType(intType()), "Create a list of numbers, every element increasing by one."),

		VSpec("cons", []VSpecInput{
			{Name: "head", Type: TVar("a")},
			{Name: "tail", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Add an element to the front of a list."),

		// Transformation
		VSpec("map", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, TVar("b"))},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("b")), "Apply a function to every element of a list."),

		VSpec("indexedMap", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{intType(), TVar("a")}, TVar("b"))},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("b")), "Same as map but the function is also applied to the index of each element."),

		VSpec("foldl", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b")}, TVar("b"))},
			{Name: "acc", Type: TVar("b")},
			{Name: "list", Type: ListType(TVar("a"))},
		}, TVar("b"), "Reduce a list from the left."),

		VSpec("foldr", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b")}, TVar("b"))},
			{Name: "acc", Type: TVar("b")},
			{Name: "list", Type: ListType(TVar("a"))},
		}, TVar("b"), "Reduce a list from the right."),

		VSpec("filter", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, boolType())},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Keep only elements that satisfy the predicate."),

		VSpec("filterMap", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, MaybeType(TVar("b")))},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("b")), "Filter and transform elements of a list."),

		// Utilities
		VSpec("length", []VSpecInput{
			{Name: "list", Type: ListType(TVar("a"))},
		}, intType(), "Determine the length of a list."),

		VSpec("reverse", []VSpecInput{
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Reverse a list."),

		VSpec("member", []VSpecInput{
			{Name: "value", Type: TVar("a")},
			{Name: "list", Type: ListType(TVar("a"))},
		}, boolType(), "Figure out whether a list contains a value."),

		VSpec("all", []VSpecInput{
			{Name: "predicate", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, boolType())},
			{Name: "list", Type: ListType(TVar("a"))},
		}, boolType(), "Determine if all elements satisfy some test."),

		VSpec("any", []VSpecInput{
			{Name: "predicate", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, boolType())},
			{Name: "list", Type: ListType(TVar("a"))},
		}, boolType(), "Determine if any elements satisfy some test."),

		VSpec("maximum", []VSpecInput{
			{Name: "list", Type: ListType(TVar("comparable"))},
		}, MaybeType(TVar("comparable")), "Find the maximum element in a non-empty list."),

		VSpec("minimum", []VSpecInput{
			{Name: "list", Type: ListType(TVar("comparable"))},
		}, MaybeType(TVar("comparable")), "Find the minimum element in a non-empty list."),

		VSpec("sum", []VSpecInput{
			{Name: "list", Type: ListType(TVar("number"))},
		}, TVar("number"), "Get the sum of the list elements."),

		VSpec("product", []VSpecInput{
			{Name: "list", Type: ListType(TVar("number"))},
		}, TVar("number"), "Get the product of the list elements."),

		// Combining
		VSpec("append", []VSpecInput{
			{Name: "list1", Type: ListType(TVar("a"))},
			{Name: "list2", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Put two lists together."),

		VSpec("concat", []VSpecInput{
			{Name: "lists", Type: ListType(ListType(TVar("a")))},
		}, ListType(TVar("a")), "Concatenate a bunch of lists into a single list."),

		VSpec("concatMap", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, ListType(TVar("b")))},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("b")), "Map a given function onto a list and flatten the results."),

		VSpec("intersperse", []VSpecInput{
			{Name: "separator", Type: TVar("a")},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Place a value between all members of a list."),

		// map2 through map5
		VSpec("map2", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b")}, TVar("c"))},
			{Name: "list1", Type: ListType(TVar("a"))},
			{Name: "list2", Type: ListType(TVar("b"))},
		}, ListType(TVar("c")), "Combine two lists, combining them with the given function."),

		VSpec("map3", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b"), TVar("c")}, TVar("d"))},
			{Name: "list1", Type: ListType(TVar("a"))},
			{Name: "list2", Type: ListType(TVar("b"))},
			{Name: "list3", Type: ListType(TVar("c"))},
		}, ListType(TVar("d")), "Combine three lists with a function."),

		VSpec("map4", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b"), TVar("c"), TVar("d")}, TVar("e"))},
			{Name: "list1", Type: ListType(TVar("a"))},
			{Name: "list2", Type: ListType(TVar("b"))},
			{Name: "list3", Type: ListType(TVar("c"))},
			{Name: "list4", Type: ListType(TVar("d"))},
		}, ListType(TVar("e")), "Combine four lists with a function."),

		VSpec("map5", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b"), TVar("c"), TVar("d"), TVar("e")}, TVar("f"))},
			{Name: "list1", Type: ListType(TVar("a"))},
			{Name: "list2", Type: ListType(TVar("b"))},
			{Name: "list3", Type: ListType(TVar("c"))},
			{Name: "list4", Type: ListType(TVar("d"))},
			{Name: "list5", Type: ListType(TVar("e"))},
		}, ListType(TVar("f")), "Combine five lists with a function."),

		// Sorting
		VSpec("sort", []VSpecInput{
			{Name: "list", Type: ListType(TVar("comparable"))},
		}, ListType(TVar("comparable")), "Sort values from lowest to highest."),

		VSpec("sortBy", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, TVar("comparable"))},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Sort values by a derived property."),

		VSpec("sortWith", []VSpecInput{
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("a")}, orderType())},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Sort values with a custom comparison function."),

		// Deconstruction
		VSpec("isEmpty", []VSpecInput{
			{Name: "list", Type: ListType(TVar("a"))},
		}, boolType(), "Determine if a list is empty."),

		VSpec("head", []VSpecInput{
			{Name: "list", Type: ListType(TVar("a"))},
		}, MaybeType(TVar("a")), "Extract the first element of a list."),

		VSpec("tail", []VSpecInput{
			{Name: "list", Type: ListType(TVar("a"))},
		}, MaybeType(ListType(TVar("a"))), "Extract the elements after the head of a list."),

		VSpec("take", []VSpecInput{
			{Name: "n", Type: intType()},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Take the first n members of a list."),

		VSpec("drop", []VSpecInput{
			{Name: "n", Type: intType()},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ListType(TVar("a")), "Drop the first n members of a list."),

		VSpec("partition", []VSpecInput{
			{Name: "predicate", Type: TFun([]ir.Type[ir.Unit]{TVar("a")}, boolType())},
			{Name: "list", Type: ListType(TVar("a"))},
		}, ir.NewTypeTuple(ir.Unit{}, []ir.Type[ir.Unit]{
			ListType(TVar("a")),
			ListType(TVar("a")),
		}), "Partition a list based on some test."),

		VSpec("unzip", []VSpecInput{
			{Name: "list", Type: ListType(ir.NewTypeTuple(ir.Unit{}, []ir.Type[ir.Unit]{
				TVar("a"),
				TVar("b"),
			}))},
		}, ir.NewTypeTuple(ir.Unit{}, []ir.Type[ir.Unit]{
			ListType(TVar("a")),
			ListType(TVar("b")),
		}), "Decompose a list of tuples into a tuple of lists."),

		// Joins
		VSpec("innerJoin", []VSpecInput{
			{Name: "right", Type: ListType(TVar("b"))},
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b")}, boolType())},
			{Name: "left", Type: ListType(TVar("a"))},
		}, ListType(ir.NewTypeTuple(ir.Unit{}, []ir.Type[ir.Unit]{
			TVar("a"),
			TVar("b"),
		})), "Inner join two lists."),

		VSpec("leftJoin", []VSpecInput{
			{Name: "right", Type: ListType(TVar("b"))},
			{Name: "f", Type: TFun([]ir.Type[ir.Unit]{TVar("a"), TVar("b")}, boolType())},
			{Name: "left", Type: ListType(TVar("a"))},
		}, ListType(ir.NewTypeTuple(ir.Unit{}, []ir.Type[ir.Unit]{
			TVar("a"),
			MaybeType(TVar("b")),
		})), "Left join two lists."),
	}
}

// ListType creates a List type reference with the given element type
func ListType(elementType ir.Type[ir.Unit]) ir.Type[ir.Unit] {
	return ir.NewTypeReference(
		ir.Unit{},
		ToFQName(ListModuleName(), "List"),
		[]ir.Type[ir.Unit]{elementType},
	)
}

// MaybeType creates a Maybe type reference (forward reference for now)
// This will be properly defined in the Maybe module
func MaybeType(valueType ir.Type[ir.Unit]) ir.Type[ir.Unit] {
	return ir.NewTypeReference(
		ir.Unit{},
		ToFQName(ir.PathFromString("Maybe"), "Maybe"),
		[]ir.Type[ir.Unit]{valueType},
	)
}
